libslide/src/emit.rs

```diff
@@ -0,0 +1,287 @@
+                        } else {
+                            child.emit_pretty()
+                        }
+                    }
+                    expr => expr.emit_pretty(),
+                };
+                result.push_str(&format!(
+                    "{} {} {}",
+                    format_arg(&self.lhs, false),
+                    self.op,
+                    format_arg(&self.rhs, true)
+                ));
+                result
+            }
+        }
+    };
+}
+display_binary_expr!(<Expr>);
+display_binary_expr!(<ExprPat>);
+
+fmt_emit_impl!(UnaryOperator);
+impl Emit for UnaryOperator {
+    fn emit_pretty(&self) -> String {
+        match self {
+            Self::SignPositive => "+",
+            Self::SignNegative => "-",
+        }
+        .to_owned()
+    }
+}
+
+macro_rules! display_unary_expr {
+    (<$expr:ident>) => {
+        fmt_emit_impl!(UnaryExpr<$expr>);
+        impl Emit for UnaryExpr<$expr> {
+            fn emit_pretty(&self) -> String {
+                let mut result = String::with_capacity(128);
+                use $expr::*;
+                let format_arg = |arg: &Rc<$expr>| match arg.as_ref() {
+                    BinaryExpr(l) => format!("({})", l),
+                    expr => expr.emit_pretty(),
+                };
+                result.push_str(&format!("{}{}", self.op, format_arg(&self.rhs)));
+                result
+            }
+        }
+    };
+}
+display_unary_expr!(<Expr>);
+display_unary_expr!(<ExprPat>);
+
+fmt_emit_impl!(ExprPat);
+impl Emit for ExprPat {
+    fn emit_pretty(&self) -> String {
+        match self {
+            Self::Const(num) => num.to_string(),
+            Self::VarPat(var) | Self::ConstPat(var) | Self::AnyPat(var) => var.to_string(),
+            Self::BinaryExpr(binary_expr) => binary_expr.emit_pretty(),
+            Self::UnaryExpr(unary_expr) => unary_expr.emit_pretty(),
+            Self::Parend(expr) => format!("({})", expr.emit_pretty()),
+            Self::Bracketed(expr) => format!("[{}]", expr.emit_pretty()),
+        }
+    }
+
+    fn emit_s_expression(&self) -> String {
+        match self {
+            Self::Const(konst) => konst.to_string(),
+            Self::VarPat(pat) | Self::ConstPat(pat) | Self::AnyPat(pat) => pat.to_string(),
+            Self::BinaryExpr(BinaryExpr { op, lhs, rhs }) => format!(
+                "({} {} {})",
+                op.to_string(),
+                lhs.emit_s_expression(),
+                rhs.emit_s_expression()
+            ),
+            Self::UnaryExpr(UnaryExpr { op, rhs }) => {
+                format!("({} {})", op.to_string(), rhs.emit_s_expression())
+            }
+            Self::Parend(inner) => format!("({})", inner.emit_s_expression()),
+            Self::Bracketed(inner) => format!("[{}]", inner.emit_s_expression()),
+        }
+    }
+}
+
+impl Emit for Rc<ExprPat> {
+    fn emit_pretty(&self) -> String {
+        self.as_ref().emit_pretty()
+    }
```

Comment:
is this a double refernce? because your pulling &self. I'm a little confused as why you need as_ref() again

Reply:
not great API, `as_ref` is to take the reference of the item held within the `Rc`